src/material/button/button-lazy-loader.ts

```diff
@@ -0,0 +1,151 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DOCUMENT} from '@angular/common';
+import {
+  ANIMATION_MODULE_TYPE,
+  ElementRef,
+  Injectable,
+  NgZone,
+  OnDestroy,
+  inject,
+} from '@angular/core';
+import {
+  MAT_RIPPLE_GLOBAL_OPTIONS,
+  MatRipple,
+  RippleConfig,
+  RippleGlobalOptions,
+  RippleRenderer,
+  RippleTarget,
+} from '@angular/material/core';
+import {Platform} from '@angular/cdk/platform';
+
+/** The options for the MatButtonRippleLoader's event listeners. */
+const OPTIONS = {capture: true};
+
+/** The events that should trigger the initialization of the ripple. */
+const RIPPLE_INTERACTION_EVENTS = ['focus', 'click', 'mouseenter', 'touchstart'];
+
+/** The attribute attached to a mat-button whose ripple has not yet been initialized. */
+export const MAT_BUTTON_RIPPLE_UNINITIALIZED = 'mat-button-ripple-uninitialized';
+
+/**
+ * Handles attaching the MatButton's ripple on demand.
+ *
+ * This service allows us to avoid eagerly creating & attaching the MatButton's ripple.
+ * It works by creating & attaching the ripple only when a MatButton is first interacted with.
+ */
+@Injectable({providedIn: 'root'})
+export class MatButtonLazyLoader implements OnDestroy {
+  private _document = inject(DOCUMENT, {optional: true});
+  private _animationMode = inject(ANIMATION_MODULE_TYPE, {optional: true});
+  private _globalRippleOptions = inject(MAT_RIPPLE_GLOBAL_OPTIONS, {optional: true});
+  private _platform = inject(Platform);
+  private _ngZone = inject(NgZone);
+
+  constructor() {
+    this._ngZone.runOutsideAngular(() => {
+      for (const event of RIPPLE_INTERACTION_EVENTS) {
+        this._document?.addEventListener(event, this._onInteraction, OPTIONS);
+      }
+    });
+  }
+
+  ngOnDestroy() {
+    for (const event of RIPPLE_INTERACTION_EVENTS) {
+      this._document?.removeEventListener(event, this._onInteraction, OPTIONS);
+    }
+  }
+
+  /** Handles creating and attaching button internals when a button is initially interacted with. */
+  private _onInteraction = (event: Event) => {
+    if (event.target === this._document) {
+      return;
+    }
+    const eventTarget = event.target as Element;
+
+    // TODO(wagnermaciel): Consider batching these events to improve runtime performance.
+
+    const button = eventTarget.closest(`[${MAT_BUTTON_RIPPLE_UNINITIALIZED}]`);
+    if (button) {
+      button.removeAttribute(MAT_BUTTON_RIPPLE_UNINITIALIZED);
+      this._appendRipple(button as HTMLButtonElement);
```

Comment:
```suggestion
      this._appendRipple(button as HTMLElement);
```
`button` isn't always an `HTMLButtonElement` because sometimes it's an anchor